### PR TITLE
afl-fuzz tests for schema parser (WIP)

### PR DESCRIPTION
**Pull request by excelangue**
Working on fixing: #7

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -4,3 +4,10 @@
 /.cabal-sandbox
 /cabal.sandbox.config
 result*
+
+# AFL stub intermediates
```
Comment (zenhack) [inline .gitignore]: Probably should just add this as another test-suite component in the cabal file, which keeps it all under dist/dist-newstyle. I wouldn't object to adding `*.o` and `*.hi` rules though.

**Comment by zenhack** [comment]
Thanks for tackling this.

**Comment by taktoa** [comment]
@excelangue
Here's a rudimentary diff for adding a Hydra job; you can modify this fairly easily to run the AFL tests.
You can save this diff to `foo.diff` and then use `git apply foo.diff ` to update your working directory with these changes.

```diff
diff --git a/Makefile b/Makefile
index 24b27ff..f90c3f5 100644
--- a/Makefile
+++ b/Makefile
@@ -25,6 +25,7 @@ help:
 	-@echo "Targets:"
 	-@echo "  configure:  cabal configure"
 	-@echo "  build:      cabal build"
+	-@echo "  check:      cabal test + any nix tests"
 	-@echo "  lib-repl:   cabal repl lib:capnp"
 	-@echo "  exe-repl:   cabal repl exe:capnpc-haskell"
 	-@echo "  test-repl:  cabal repl test:the-test-suite"
@@ -38,6 +39,10 @@ configure:
 build: configure
 	cabal build -j
 
+check: build
+	cabal test -j
+	nix-build --no-out-link release.nix -A capnp-afl
+
 lib-repl: configure
 	cabal repl "lib:capnp"
 
diff --git a/release.nix b/release.nix
index a8b27a6..f5c1c4f 100644
--- a/release.nix
+++ b/release.nix
@@ -68,4 +68,30 @@ with rec {
 
 {
   capnp = addHydraHaddock haskellPackages haskellPackages.capnp;
+
+  capnp-afl = pkgs.stdenv.mkDerivation {
+    name = "capnp-afl";
+
+    src = haskellPackages.capnp.src;
+
+    buildInputs = [
+      (haskellPackages.ghcWithPackages (p: [ p.capnp ]))
+    ];
+
+    doCheck = true;
+
+    phases = ["unpackPhase" "buildPhase" "checkPhase"];
+
+    buildPhase = ''
+      cd tests/afl-fuzz
+      ghc -c -O2 Decode.hs
+      ghc --make  -no-hs-main -optc-O main.c Decode -o main
+      rm Decode.o Decode.hi
+    '';
+
+    checkPhase = ''
+      mkdir -p $out
+      cat ../data/schema-codegenreq | ./main > $out/output.log
+    '';
+  };
 }
```

**Comment by zenhack** [comment]
Re: the runtime dependency, our existing test suite already fails that metric, since it calls the `capnp` command line tool. The fact that you can't specify dependencies on executables (even ones packaged with cabal) is awkward, but not a reason to just abandon the build system entirely.

As far as getting cabal to run the test under afl, you could e.g. check an environment variable, and depending on the value, either run the test stuff or change the env var and then call afl, passing the path to the test executable.

Also, looking at this a little more carefully, it doesn't seem like having main be in C actually does anything for us: If we're not going to try to use afl's instrumentation, we can just pass `-n` to do dumb mode. If we are, just having a trivial main function instrumented isn't going to do much for us -- it would only make any sense if we're going to use ghc's C backend and then run that through afl-gcc (in which case, there's no need to write a main in C either).

I don't think AFL has been used on haskell programs very much -- this PR is the third result I got on ddg, and the earlier ones aren't really relevant. I don't know whether trying to use the instrumentation is going to even be useful.

**Comment by zenhack** [comment]
Worth noting, when I mentioned AFL in the original issue, I hadn't thought too deeply about it, and had forgotten that you needed instrumentation for the fancy stuff. Without that, it's not obvious to me that we're getting anything we couldn't do with quickcheck.

**Comment by taktoa** [comment]
Good to know. I'm meeting with @excelangue later today so I'll talk to him about switching to QuickCheck.

**Comment by excelangue** [comment]
Will be moving to QuickCheck.

**Pull request closed without merging**